### PR TITLE
Implement CUDA-based gamma sampling

### PR DESCRIPTION
**Pull request by rachtsingh**
I've taken a stab at reducing code duplication between the CUDA/CPU implementations using one strategy, but happy to try something else if there's an alternative.

### CODE REVIEW

**Comment by apaszke** [comment]
@rachtsingh I think you have accidentally updated the subtrees

**Comment by rachtsingh** [comment]
Yes, sorry, just fixed (I think my understanding of subtrees is still a little incomplete).

**Comment by rachtsingh** [comment]
Hmm, there are test failures (due to `Distributions.cuh` not having `curand.h` linked correctly, I think), but I can't reproduce it on my devbox. I don't know much about how the linking happens - can anyone help?

**Comment by rachtsingh** [comment]
Ok, I think it's ready for review - don't love the code dedup solution but very open to alternatives. Maybe @apaszke ?

File: test/test_distributions.py [hunk 1]
```diff
@@ -599,7 +599,7 @@ def test_poisson_sample(self):
     @unittest.skipIf(not TEST_CUDA, "CUDA not found")
     @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
     def test_poisson_gpu_sample(self):
-        set_rng_seed(0)
+        set_rng_seed(1)
```
Comment (rachtsingh) [inline test/test_distributions.py]: This is an unfortunate change - I tried a few different rng seed values to see what was going on, but it looks like it *is* one of the 'randomized errors', and `0` is the only bad value. Not happy but I'll keep a close eye.

**Comment by gchanan** [comment]
can you manually check that `test_distributions.py` still works when you compile with `WITH_SCALARS=1`?

**Comment by rachtsingh** [comment]
Sure, running the build now.

**Comment by rachtsingh** [comment]
Hmm, it fails with 
```
torch/csrc/autograd/generated/VariableType.cpp:9775:30: error: ‘ensure_no_aten_scalars’ was not declared in this scope
```
Do I need to `python setup.py clean` first?

**Comment by rachtsingh** [comment]
Ok, it works just fine after I cleaned and rebuilt with `WITH_SCALARS=1`.

**Comment by rachtsingh** [comment]
this is no longer ready for review, assuming the changes in this PR: https://github.com/pytorch/pytorch/pull/4892 affect it. I'll revisit this later today and see what changes need to be made.

**Comment by rachtsingh** [comment]
Closing for now to avoid retriggering spurious builds (and because it doesn't work as-is)

**Pull request closed without merging**